### PR TITLE
[Refactoring] Rely on Java-8 for implementing 

### PR DESCRIPTION
**Pull request by chibenwa**

### CODE REVIEW

File: mailbox/api/src/main/java/org/apache/james/mailbox/model/MailboxMetaData.java [hunk 1]
```diff
@@ -65,6 +64,10 @@ public static MailboxMetaData unselectableMailbox(MailboxPath path, MailboxId ma
         return new MailboxMetaData(path, mailboxId, delimiter, Children.CHILDREN_ALLOWED_BUT_UNKNOWN, Selectability.NONE);
     }
 
+    public static  final Comparator<MailboxMetaData> COMARATOR = Comparator
```
Comment (Arsnael) [inline mailbox/api/src/main/java/org/apache/james/mailbox/model/MailboxMetaData.java]: s/COMARATOR/COMPARATOR

File: mailbox/store/src/test/java/org/apache/james/mailbox/store/search/CombinedComparatorTest.java [hunk 2]
```diff
@@ -123,12 +125,18 @@ public void createShouldConvertDisplayFrom() {
...
-        assertThat(CombinedComparator.create(ImmutableList.of(new Sort(SortClause.DisplayFrom, Order.REVERSE))).getComparators())
-            .containsOnly(new ReverseComparator(HeaderDisplayComparator.FROM_COMPARATOR));
+        MailboxMessage message1 = mock(MailboxMessage.class);
+        when(message1.getUid()).thenReturn(MessageUid.of(1));
+        MailboxMessage message2 = mock(MailboxMessage.class);
+        when(message2.getUid()).thenReturn(MessageUid.of(2));
+
+        Comparator<MailboxMessage> comparator = CombinedComparator.create(ImmutableList.of(new Sort(SortClause.Uid, Order.REVERSE)));
+
+        assertThat(comparator.compare(message1, message2)).isGreaterThan(0);
```
Comment (rouazana) [inline mailbox/store/src/test/java/org/apache/james/mailbox/store/search/CombinedComparatorTest.java]: why this change?

Comment (chibenwa) [reply mailbox/store/src/test/java/org/apache/james/mailbox/store/search/CombinedComparatorTest.java]: Because we can no longer assert that we are using a reversed comparator. The best way to check that it is reversed is actually to execute it.

File: mailbox/jpa/src/main/java/org/apache/james/mailbox/jpa/mail/model/openjpa/AbstractJPAMailboxMessage.java [hunk 3]
```diff
@@ -98,7 +97,7 @@
 @MappedSuperclass
 public abstract class AbstractJPAMailboxMessage implements MailboxMessage {
 
-    private static final Comparator<MailboxMessage> MESSAGE_UID_COMPARATOR = new UidComparator();
+    private static final Comparator<MailboxMessage> MESSAGE_UID_COMPARATOR = Comparator.comparing(MailboxMessage::getUid);
     private static final String TOSTRING_SEPARATOR = " ";
```
Comment (remk) [inline mailbox/jpa/src/main/java/org/apache/james/mailbox/jpa/mail/model/openjpa/AbstractJPAMailboxMessage.java]: not relatated to this PR but s/TOSTRING_SEPARATOR/TO_STRING_SEPARATOR

**Comment by chibenwa** [comment]
test this please

**Comment by chibenwa** [comment]
test this plesse

**Comment by chibenwa** [comment]
test this please

**Comment by Arsnael** [comment]
```
 [e7be7a4e877bc65a94863549d1878069229707b7] [ERROR]   CassandraRenameTest>Rename.testRenameHierarchyIT:88 » InvalidServerResponse 
[e7be7a4e877bc65a94863549d1878069229707b7] L...
[e7be7a4e877bc65a94863549d1878069229707b7] [ERROR]   CassandraRenameTest>Rename.testRenameHierarchyKO:81 » InvalidServerResponse 
[e7be7a4e877bc65a94863549d1878069229707b7] L...
[e7be7a4e877bc65a94863549d1878069229707b7] [ERROR]   CassandraRenameTest>Rename.testRenameHierarchyUS:74 » InvalidServerResponse 
[e7be7a4e877bc65a94863549d1878069229707b7] L...
```

**Comment by chibenwa** [comment]
`EventDeadLettersIntegrationTest.failedEventShouldNotBeInDeadLettersAfterSuccessfulRedelivery:332->retrieveFirstFailedInsertionId:195 » IndexOutOfBounds`

test this please

**Comment by chibenwa** [comment]
`DefaultMailboxesProvisioningFilterTest.createMailboxesIfNeededShouldNotGenerateExceptionsInConcurrentEnvironment:94 » Execution`

test this please

**Comment by chibenwa** [comment]
test this please

File: mailbox/api/src/main/java/org/apache/james/mailbox/model/MailboxMetaData.java [hunk 4]
```diff
@@ -65,6 +64,10 @@ public static MailboxMetaData unselectableMailbox(MailboxPath path, MailboxId ma
...
+        .<MailboxMetaData, Boolean>comparing(metadata -> metadata.getPath().isInbox())
```
Comment (mbaechler) [inline mailbox/api/src/main/java/org/apache/james/mailbox/model/MailboxMetaData.java]: please provide evidences that the code you rewrite is well covered by tests with every commit

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/search/comparator/CombinedComparator.java [hunk 5]
```diff
@@ -51,35 +51,35 @@ public static CombinedComparator create(List<Sort> sorts) {
     private static Comparator<MailboxMessage> toComparator(Sort sort) {
```
Comment (chibenwa) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/search/comparator/CombinedComparator.java]: Easy search comparator are covered by sorts of Scrolling search testing, which is running our pretty complete search test suite.

File: mailbox/api/src/main/java/org/apache/james/mailbox/StandardMailboxMetaDataComparator.java [hunk 6]
```diff
@@ -1,66 +0,0 @@
-/****************************************************************
- * Licensed to the Apache Software Foundation (ASF) under one   *
```
Comment (chibenwa) [inline mailbox/api/src/main/java/org/apache/james/mailbox/StandardMailboxMetaDataComparator.java]: Mailbox ordering testing in LIST answer is tested by our MPT IMAP test suite.

Not ideal but enough to be safe.

File: server/dns-service/dnsservice-dnsjava/src/main/java/org/apache/james/dnsservice/dnsjava/DNSJavaService.java [hunk 7]
```diff
@@ -106,8 +106,11 @@
 
     /**
      * The MX Comparator used in the MX sort.
+     *
+     * RFC 2821 section 5 requires that we sort the MX records by their
+     * preference.
      */
-    private final Comparator<MXRecord> mxComparator = new MXRecordComparator();
+    private final Comparator<MXRecord> mxComparator = Comparator.comparing(MXRecord::getPriority);
```
Comment (chibenwa) [inline server/dns-service/dnsservice-dnsjava/src/main/java/org/apache/james/dnsservice/dnsjava/DNSJavaService.java]: MX record ordering is tested by our new FakeSMTP RemoteDelivery tests

File: mailbox/jpa/src/main/java/org/apache/james/mailbox/jpa/mail/model/openjpa/AbstractJPAMailboxMessage.java [hunk 8]
```diff
@@ -59,7 +59,6 @@
 import org.apache.james.mailbox.store.mail.model.Property;
```
Comment (chibenwa) [inline mailbox/jpa/src/main/java/org/apache/james/mailbox/jpa/mail/model/openjpa/AbstractJPAMailboxMessage.java]: Not sure where that one is used though....

I think we can rely on default method to avoid getting this "comparable" behaviour implementation specific....

**Comment by chibenwa** [comment]
test this please

**Comment by chibenwa** [comment]
`[4965d5c343b4b6c71a29cf7d31d822033d07938f] [ERROR] Failed to execute goal on project james-server-lifecycle-api: Could not resolve dependencies for project org.apache.james:james-server-lifecycle-api:jar:3.5.0-SNAPSHOT: Failed to collect dependencies at com.google.guava:guava:jar:25.1-jre -> org.codehaus.mojo:animal-sniffer-annotations:jar:1.14: Failed to read artifact descriptor for org.codehaus.mojo:animal-sniffer-annotations:jar:1.14: Could not transfer artifact org.codehaus:codehaus-parent:pom:4 from/to central (https://repo.maven.apache.org/maven2): /root/.m2/repository/org/codehaus/codehaus-parent/4/codehaus-parent-4.pom.part (No such file or directory) -> [Help 1]`

test this please

**Comment by chibenwa** [comment]
Merged
